lib/Serialization/Serialization.cpp

```diff
@@ -2848,13 +2808,6 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
     for (auto Attr : D->getAttrs())
       writeDeclAttribute(Attr);
 
-    if (auto VD = dyn_cast<ValueDecl>(D)) {
-      // Hack: synthesize a 'final' attribute if finality was inferred.
-      if (VD->isFinal() && !D->getAttrs().hasAttribute<FinalAttr>())
```

Comment:
Has this become impossible?

Reply:
Would it be worthwhile to have an assertion here to double-check that `FinalAttr` is in sync or is that not useful because correctness is guaranteed by construction?

Reply:
Good point. I'll add it to the AST verifier; that's where such correct-by-construction assertions belong.